consensus/src/consensus/inner.rs

```diff
@@ -0,0 +1,508 @@
+
+        while let Some(message) = receiver.recv().await {
+            match message {
+                ConsensusMessage::ReceiveTransaction(transaction, response) => {
+                    response.send(self.receive_transaction(transaction)).ok();
+                }
+                ConsensusMessage::VerifyTransaction(transaction, response) => {
+                    let out = match self.verify_transaction(&*transaction) {
+                        Ok(out) => out,
+                        Err(e) => {
+                            error!(
+                                "failed to validate transaction -- note this does not mean the transaction was valid or invalid: {:?}",
+                                e
+                            );
+                            false
+                        }
+                    };
+                    response.send(out).ok();
+                }
+                ConsensusMessage::ReceiveBlock(block, response) => match self.receive_block(&block).await {
+                    Ok(()) => {
+                        response.send(true).ok();
+                    }
+                    Err(e) => {
+                        error!("failed receiving block: {:?}", e);
+                        response.send(false).ok();
+                    }
+                },
+                ConsensusMessage::FetchMemoryPool(size, response) => {
+                    let out = self
+                        .memory_pool
+                        .get_candidates(size)
+                        .map(|x| x.into_iter().cloned().collect());
+                    response.send(out).ok();
+                }
+                ConsensusMessage::CreateTransaction(request, response) => {
+                    let out = self.create_transaction(*request);
+                    response.send(out).ok();
+                }
+                ConsensusMessage::CreatePartialTransaction(request, response) => {
+                    let out = self.create_partial_transaction(request);
+                    response.send(out).ok();
+                }
+            }
+        }
+    }
+
+    /// Check if the transaction is valid.
+    fn verify_transaction(&self, transaction: &SerialTransaction) -> Result<bool, ConsensusError> {
+        if !self
+            .public
+            .parameters
+            .authorized_inner_snark_ids
+            .contains(&to_bytes![transaction.inner_circuit_id]?)
+        {
+            return Ok(false);
+        }
+
+        Ok(InstantiatedDPC::verify(
+            &self.public.public_parameters,
+            &Tx::deserialize(transaction)?,
+            &self.ledger.deserialize::<Components>(),
+        )?)
+    }
+
+    /// Check if the transactions are valid.
+    fn verify_transactions(&self, transactions: &[SerialTransaction]) -> Result<bool, ConsensusError> {
+        for tx in transactions {
+            if !self
+                .public
+                .parameters
+                .authorized_inner_snark_ids
```

Comment:
We might want to change this into a `HashSet` for faster lookups, unless there's very few of them; other than this, instead of manual iteration, couldn't this use `contains`?

Reply:
`contains` doesn't work because it doesn't handle AsRef stuff (i.e. `Vec<u8>` asref -> `&[u8]`). using `any` is a common pattern.